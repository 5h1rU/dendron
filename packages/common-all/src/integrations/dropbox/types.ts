import { files } from "dropbox";

export type MetadataReference =
  | files.FileMetadataReference
  | files.FolderMetadataReference;

export interface ListFolderResultSimple {
  /**
   * The files and (direct) subfolders in the folder.
   */
  entries: Array<MetadataReference>;
  /**
   * Pass the cursor into listFolderContinue() to see what's changed in the
   * folder since your previous query.
   */
  cursor: files.ListFolderCursor;
  /**
   * If true, then there are more entries available. Pass the cursor to
   * listFolderContinue() to retrieve the rest.
   */
  has_more: boolean;
}
