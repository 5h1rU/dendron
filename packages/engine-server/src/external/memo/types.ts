import { URI } from "vscode-uri";

export type WorkspaceCache = {
  imageUris: URI[];
  markdownUris: URI[];
  otherUris: URI[];
  allUris: URI[];
  danglingRefs: string[];
  danglingRefsByFsPath: { [key: string]: string[] };
};

export type RefT = {
  label: string;
  ref: string;
};
