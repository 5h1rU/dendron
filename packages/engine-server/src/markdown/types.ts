import { DNoteRefLink, NotePropsV2 } from "@dendronhq/common-all";
import MDAST from "mdast";

export type DendronASTNode = MDAST.Parent & {
  notes?: NotePropsV2[];
  children: MDAST.Parent["children"] | DendronASTNode[];
};

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  PARAGRAPH = "paragraph",
}

export enum DendronASTDest {
  // MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  HTML = "HTML",
}

export type DendronASTData = {
  dest: DendronASTDest;
};

export type WikiLinkNoteV4 = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  value: string;
  data: WikiLinkDataV4;
};

export type WikiLinkDataV4 = {
  alias: string;
  anchorHeader?: string;
  prefix?: string;
};

export type NoteRefNoteV4 = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  value: string;
  data: NoteRefDataV4;
};

export type NoteRefDataV4 = {
  link: DNoteRefLink;
};
