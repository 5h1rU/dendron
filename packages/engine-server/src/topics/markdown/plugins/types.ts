import { DNoteLoc, DNoteRefLink, NotePropsV2 } from "@dendronhq/common-all";
import MDAST from "mdast";

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  PARAGRAPH = "paragraph",
}

export enum DendronASTDest {
  MD_ENHANCED_PREVIEW,
  MD_REGULAR,
  HTML,
}

// --- General

export type DendronASTData = {
  dest: DendronASTDest;
};

// @deprecate
export type LinkOpts = {
  toMdEnhancedPreview?: boolean;
};

export type DendronASTNode = MDAST.Parent & {
  notes?: NotePropsV2[];
};

// --- WIKI Link
export type RefLinkNote = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  data: RefLinkData;
  value: string;
};

export type RefLinkData = {
  link: DNoteRefLink;
  // --- Old
  alias: string;
  permalink: string;
  exists: boolean;
  hName: string;
  hProperties: any;
  hChildren: any[];
  toMd?: boolean;
  prefix?: string;
  useId: boolean;
};

export type WikiLinkNote = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  data: WikiLinkData;
  value: string;
};

export type WikiLinkData = {
  alias: string;
  permalink: string;
  exists: boolean;
  hName: string;
  hProperties: {
    className: string;
    href: string;
  };
  hChildren: any[];
  anchorHeader?: string;

  // dendron specific
  toMd?: boolean;
  toHTML?: boolean;
  /**
   * {ROOT_URL}/{prefix?}/{destination}
   */
  prefix?: string;
  useId: boolean;
  note?: { id: string };
  replace?: DNoteLoc;
  forNoteRefInPreview?: boolean;
  forNoteRefInSite?: boolean;
} & LinkOpts;
