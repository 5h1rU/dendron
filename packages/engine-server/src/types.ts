import { DEngineClientV2, DNoteLoc } from "@dendronhq/common-all";

export type ReplaceLinkOpts = { from: DNoteLoc; to: DNoteLoc };
export { DEngineClientV2 };
