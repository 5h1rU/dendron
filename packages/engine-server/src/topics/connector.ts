import {
  DendronConfig,
  DendronError,
  DEngineClientV2,
  Time,
} from "@dendronhq/common-all";
import { createFileWatcher, DLogger } from "@dendronhq/common-server";
import fs, { FSWatcher } from "fs-extra";
import _ from "lodash";
import { DConfig } from "../config";
import { DendronEngineClient } from "../engineClient";
import {
  getPortFilePath,
  getWSMetaFilePath,
  openPortFile,
  openWSMetaFile,
} from "../utils";

export type EngineConnectorInitOpts = {
  onReady?: ({}: { ws: EngineConnector }) => Promise<void>;
  numRetries?: number;
  portOverride?: number;
};

export class EngineConnector {
  public wsRoot: string;
  //public vaults: DVault[];
  public _engine: DEngineClientV2 | undefined;
  public port: number | undefined;
  public onReady?: ({ ws }: { ws: EngineConnector }) => Promise<void>;
  public serverPortWatcher?: FSWatcher;
  public initialized: boolean;
  public config: DendronConfig;
  public logger?: DLogger;

  static _ENGINE_CONNECTOR: EngineConnector | undefined;

  static instance() {
    if (!this._ENGINE_CONNECTOR) {
      throw new DendronError({ msg: "no workspace" });
    }
    return this._ENGINE_CONNECTOR;
  }

  static getOrCreate({ wsRoot, logger }: { wsRoot: string; logger?: DLogger }) {
    if (!this._ENGINE_CONNECTOR) {
      return new EngineConnector({ wsRoot, logger });
    }
    return this._ENGINE_CONNECTOR;
  }

  constructor({ wsRoot, logger }: { wsRoot: string; logger?: DLogger }) {
    this.wsRoot = wsRoot;
    this.logger = logger;
    this.config = DConfig.getOrCreate(wsRoot);
    EngineConnector._ENGINE_CONNECTOR = this;
    this.initialized = false;
  }

  get vaults(): string[] {
    return this.config.vaults.map((ent) => ent.fsPath);
  }

  async init(opts?: EngineConnectorInitOpts) {
    // init engine
    this.onReady = opts?.onReady;
    if (opts?.portOverride) {
      await this.initEngine({ port: opts.portOverride });
    } else {
      return this.createServerWatcher({ numRetries: opts?.numRetries });
    }
  }

  async initEngine({ port }: { port: number }) {
    const { wsRoot, vaults } = this;
    const dendronEngine = DendronEngineClient.create({
      port,
      ws: wsRoot,
      vaults: vaults.map((ent) => ent),
      logger: this.logger,
    });
    await dendronEngine.sync();
    this._engine = dendronEngine;
    this.initialized = true;
    return dendronEngine;
  }

  get engine(): DEngineClientV2 {
    if (!this._engine) {
      throw Error("engine not set");
    }
    return this._engine;
  }

  async createServerWatcher(opts?: { numRetries?: number }) {
    const { wsRoot } = this;
    const fpath = getPortFilePath({ wsRoot });
    const { watcher } = await createFileWatcher({
      fpath,
      numTries: opts?.numRetries,
      onChange: async ({ fpath }) => {
        const port = openPortFile({ fpath });
        this.onChangePort({ port });
      },
      onCreate: async ({ fpath }) => {
        const port = openPortFile({ fpath });
        this.onChangePort({ port });
      },
    });
    // file should exist at this point
    const metaFpath = getWSMetaFilePath({ wsRoot });
    const wsMeta = openWSMetaFile({ fpath: metaFpath });
    const wsActivation = wsMeta.activationTime;

    // check if port was created after current ws
    const portCreated = Time.DateTime.fromJSDate(
      fs.statSync(fpath).ctime
    ).toMillis();
    if (portCreated > wsActivation) {
      const port = openPortFile({ fpath });
      this.onChangePort({ port });
    }
    // race condition were old workspace file is found
    setTimeout(() => {
      if (fs.existsSync(fpath) && portCreated > wsActivation) {
        const port = openPortFile({ fpath });
        this.onChangePort({ port });
      }
    }, 10000);

    // attach watcher
    this.serverPortWatcher = watcher;
  }

  async onChangePort({ port }: { port: number }) {
    const portPrev = this.port;
    if (this.port !== port) {
      this.port = port;
      await this.initEngine({ port });
    }
    if (_.isUndefined(portPrev) && this.onReady) {
      this.onReady({ ws: this });
    }
    this.initialized = true;
  }
}
