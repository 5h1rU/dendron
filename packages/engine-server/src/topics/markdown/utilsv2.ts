import {
  DLink,
  DNoteLoc,
  NotePropsV2,
  NoteUtilsV2,
} from "@dendronhq/common-all";
import _ from "lodash";
import remark from "remark";
import frontmatterPlugin from "remark-frontmatter";
import markdownParse from "remark-parse";
import {
  DendronLinksOpts,
  dendronLinksPlugin,
  WikiLinkNote,
} from "./plugins/dendronLinksPlugin";
const selectAll = require("unist-util-select").selectAll;

export class ParserUtilsV2 {
  static createWikiLinkRE(opts?: { oldLink: string }) {
    const { oldLink } = opts || {};
    if (oldLink) {
      const match = ParserUtilsV2.escapeForRegExp(oldLink);
      return `\\[\\[\\s*?(.*\\|)?\\s*${match}\\s*\\]\\]`;
    }
    return "\\[\\[\\s*?(.*\\|)?\\s*(?<name>.*)\\s*\\]\\]";
  }

  static getRemark(opts?: { dendronLinksOpts: DendronLinksOpts }) {
    const { dendronLinksOpts } = _.defaults(opts, { dendronLinksOpts: {} });
    return remark()
      .use(markdownParse, { gfm: true })
      .use(frontmatterPlugin, ["yaml"])
      .use(dendronLinksPlugin, dendronLinksOpts)
      .use({ settings: { listItemIndent: "1", fences: true } });
  }

  // static findLinks({ note }: { note: NotePropsV2 }): DLink[] {
  //   const content = note.body;
  //   const reWiki = ParserUtilsV2.createWikiLinkRE();
  //   const wikiLinks = content.matchAll(new RegExp(reWiki, "gi"));
  //   const wikiLinksMatches = Array.from(wikiLinks, (m) => {
  //     if (_.isUndefined(m.index)) {
  //       throw Error("no index found, findLinks");
  //     }
  //     // will barf at `[[]]` case
  //     if (!m.groups?.name) {
  //       throw Error(`no name found, findLInks, ${JSON.stringify(note)}`);
  //     }
  //     return {
  //       match: m[0],
  //       value: m.groups.name,
  //       alias: m.groups?.alias,
  //       start: m.index,
  //       end: m.index + m[0].length,
  //     };
  //   });
  //   return wikiLinksMatches.map((m) => ({
  //     type: "wiki",
  //     from: NoteUtilsV2.toLoc(note),
  //     original: m.match,
  //     value: m.value,
  //     alias: m?.alias,
  //     pos: { start: m.start, end: m.end },
  //     to: {
  //       fname: m.value,
  //     },
  //   }));
  // }

  static findLinks({ note }: { note: NotePropsV2 }): DLink[] {
    const content = note.body;
    let remark = ParserUtilsV2.getRemark();
    //const out = remark.processSync(content);
    let out = remark.parse(content);
    let out2: WikiLinkNote[] = selectAll("wikiLink", out);
    const dlinks = out2.map(
      (m: WikiLinkNote) =>
        ({
          type: "wiki",
          from: NoteUtilsV2.toLoc(note),
          original: m.value,
          value: m.value,
          alias: m.data.alias,
          pos: { start: m.position?.start.offset, end: m.position?.end.offset },
          to: {
            fname: m.value,
          },
        } as DLink)
    );
    return dlinks as DLink[];
  }

  static escapeForRegExp(value: string) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  }

  static async replaceLinks(opts: {
    content: string;
    from: DNoteLoc;
    to: DNoteLoc;
  }) {
    const { content, from, to } = opts;
    let remark = ParserUtilsV2.getRemark({
      dendronLinksOpts: {
        replaceLink: { from, to },
      },
    });
    const out = await remark.process(content);
    return out.toString();
  }
}
