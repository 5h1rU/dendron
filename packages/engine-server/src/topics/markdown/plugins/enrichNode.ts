import { DendronError } from "@dendronhq/common-all";
import { NoteUtilsV2 } from "@dendronhq/common-all";
import _ from "lodash";
import Unified, { Transformer } from "unified";
import visit from "unist-util-visit";
import { VFile } from "vfile";
import { EngineConnector } from "../../connector";
import {
  DendronASTData,
  DendronASTDest,
  DendronASTTypes,
  WikiLinkNote,
} from "./types";

function wikiNode2HTML(node: WikiLinkNote) {
  if (_.isUndefined(node.notes)) {
    return new DendronError({ msg: `no note found for node: ${node}` });
  }
  if (node.notes.length > 1) {
    return new DendronError({ msg: `multiple notes found for link: ${node}` });
  }
  if (node.notes.length < 1) {
    return new DendronError({
      msg: `no notes found for link: ${JSON.stringify(node)}`,
    });
  }
  const note = node.notes[0];
  const href = `${node.data.prefix || ""}${note.id}.html`;
  node.data.hProperties.href = href;
  return;
}

export function enrichNode(this: Unified.Processor): Transformer {
  return async (tree, _file: VFile) => {
    const proc = this;
    // const engine = proc.data("engine") as EngineConnector;
    const data = proc.data("dendron") as DendronASTData;

    const engine = EngineConnector.instance();
    if (!engine.initialized) {
      console.log("init engine");
      await engine.init({ portOverride: 3006 });
      proc().data("engine", engine);
    }

    visit(tree, DendronASTTypes.WIKI_LINK, (node: WikiLinkNote) => {
      // add data to node
      const fname = NoteUtilsV2.normalizeFname(node.value);
      const notes = NoteUtilsV2.getNotesByFname({
        fname,
        notes: engine.engine.notes,
      });
      node.notes = notes;
      // transform note for HTML
      switch (data.dest) {
        case DendronASTDest.HTML: {
          const err = wikiNode2HTML(node);
          if (err) {
            const errors = proc.data("errors") as DendronError[];
            errors.push(err);
            proc().data("errors", errors);
          }
          break;
        }
        default:
          // DO NOTHING
          break;
      }
      return true;
    });
  };
}
