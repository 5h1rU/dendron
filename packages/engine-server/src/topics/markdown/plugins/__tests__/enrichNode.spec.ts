import { EngineConnector } from "../../../connector";
import { ParserUtilsV2 } from "../../utilsv2";
import { enrichNode } from "../enrichNode";
import { DendronASTData, DendronASTDest } from "../types";
const remarkRehype = require("remark-rehype");
const rehypeStringify = require("rehype-stringify");

function getProc() {
  const ec = new EngineConnector({
    wsRoot: "/Users/kevinlin/projects/dendronv2/dendron-site",
  });
  const proc = ParserUtilsV2.getRemark()
    .data("engine", ec)
    .data("dendron", { dest: DendronASTDest.HTML } as DendronASTData)
    .use(enrichNode);
  return proc;
}

describe("basic", () => {
  test.only("init", async () => {
    const testString = "[[foo.md]]";
    const resp = getProc().parse(testString);
    expect(resp).toMatchSnapshot("raw");

    const resp2 = await getProc()
      .use(remarkRehype)
      .use(rehypeStringify)
      .process(testString);
    expect(resp2).toMatchSnapshot("compiled");
    expect(
      resp2.contents
        .toString()
        .indexOf("45a154ad-4b5e-4dd3-bdf6-e24a3a60ca68.html") >= 0
    ).toBeTruthy();
  });
});
