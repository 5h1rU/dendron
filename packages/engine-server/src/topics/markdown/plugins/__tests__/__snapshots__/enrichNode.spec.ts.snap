// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic init: compiled 1`] = `
VFile {
  "contents": "<p><a class=\\"internal new\\" href=\\"45a154ad-4b5e-4dd3-bdf6-e24a3a60ca68.html\\">foo.md</a></p>",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`basic init: raw 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "foo.md",
            "anchorHeader": undefined,
            "exists": false,
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "foo.md",
              },
            ],
            "hName": "a",
            "hProperties": Object {
              "className": "internal new",
              "href": "#/page/foo",
            },
            "permalink": "foo",
            "toMdEnhancedPreview": undefined,
          },
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "wikiLink",
          "value": "foo.md",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 1,
          "offset": 10,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 1,
      "offset": 10,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
